Updates to the S3 source documentation (#6379)
opensearch-project/documentation-website: _data-prepper/pipelines/configuration/sources/s3.md

# s3 source

`s3` is a source plugin that reads events from [Amazon Simple Storage Service (Amazon S3)](https://aws.amazon.com/s3/) objects. You can configure the source to either use an [Amazon Simple Queue Service (Amazon SQS)](https://aws.amazon.com/sqs/) queue or scan an S3 bucket:

- To use Amazon SQS notifications, configure S3 event notifications on your S3 bucket. After Amazon SQS is configured, the `s3` source receives messages from Amazon SQS. When the SQS message indicates that an S3 object has been created, the `s3` source loads the S3 objects and then parses them using the configured [codec](#codec).

> - To use an S3 bucket, configure the `s3` source to use Amazon S3 Selectinstead of Data Prepper to parse S3 objects.

Reviewer: ```suggestion
- To use an S3 bucket, configure the `s3` source to use Amazon S3 Select instead of Data Prepper to parse S3 objects.
```

# s3 source

`s3` is a source plugin that reads events from [Amazon Simple Storage Service (Amazon S3)](https://aws.amazon.com/s3/) objects.
You can configure the source to either use an [Amazon Simple Queue Service (Amazon SQS)](https://aws.amazon.com/sqs/) queue or scan an S3 bucket:

> - To use Amazon SQS notifications,  configure [S3 event notifications](https://docs.aws.amazon.com/AmazonS3/latest/userguide/NotificationHowTo.html) on your S3 bucket. After Amazon SQS is configured, the `s3` source receives messages from Amazon SQS. When the SQS message indicates that an S3 object was created, the `s3` source loads the S3 objects and then parses them using the configured [codec](#codec). 

Reviewer: ```suggestion
- To use Amazon SQS notifications, configure [S3 event notifications](https://docs.aws.amazon.com/AmazonS3/latest/userguide/NotificationHowTo.html) on your S3 bucket. After Amazon SQS is configured, the `s3` source receives messages from Amazon SQS. When the SQS message indicates that an S3 object was created, the `s3` source loads the S3 objects and then parses them using the configured [codec](#codec). 
```

# s3 source

`s3` is a source plugin that reads events from [Amazon Simple Storage Service (Amazon S3)](https://aws.amazon.com/s3/) objects. You can configure the source to either use an [Amazon Simple Queue Service (Amazon SQS)](https://aws.amazon.com/sqs/) queue or scan an S3 bucket:

> - To use Amazon SQS notifications, configure [S3 event notifications](https://docs.aws.amazon.com/AmazonS3/latest/userguide/NotificationHowTo.html) on your S3 bucket. After Amazon SQS is configured, the `s3` source receives messages from Amazon SQS. When the SQS message indicates that an S3 object was created, the `s3` source loads the S3 objects and then parses them using the configured [codec](#codec).

Reviewer 2: ```suggestion
- To use Amazon SQS notifications, configure [S3 event notifications](https://docs.aws.amazon.com/AmazonS3/latest/userguide/NotificationHowTo.html) on your S3 bucket. After Amazon SQS is configured, the `s3` source receives messages from Amazon SQS. When the SQS message indicates that an S3 object has been created, the `s3` source loads the S3 objects and then parses them using the configured [codec](#codec).
```

# s3 source

`s3` is a source plugin that reads events from [Amazon Simple Storage Service (Amazon S3)](https://aws.amazon.com/s3/) objects.

> You can configure the source to either use an [Amazon Simple Queue Service (Amazon SQS)](https://aws.amazon.com/sqs/) queue or scan an S3 bucket.

Reviewer: ```suggestion
You can configure the source to either use an [Amazon Simple Queue Service (Amazon SQS)](https://aws.amazon.com/sqs/) queue or scan an S3 bucket:

```

> To use Amazon SQS notifications, you configure [S3 Event Notifications](https://docs.aws.amazon.com/AmazonS3/latest/userguide/NotificationHowTo.html) on your S3 bucket. After Amazon SQS is configured, the `s3` source receives messages from Amazon SQS. When the SQS message indicates that an S3 object was created, the `s3` source loads the S3 objects and then parses them using the configured [codec](#codec). You can also configure the `s3` source to use [Amazon S3 Select](https://docs.aws.amazon.com/AmazonS3/latest/userguide/selecting-content-from-objects.html) instead of Data Prepper to parse S3 objects.

Reviewer: ```suggestion
- To use Amazon SQS notifications,  configure [S3 event notifications](https://docs.aws.amazon.com/AmazonS3/latest/userguide/NotificationHowTo.html) on your S3 bucket. After Amazon SQS is configured, the `s3` source receives messages from Amazon SQS. When the SQS message indicates that an S3 object was created, the `s3` source loads the S3 objects and then parses them using the configured [codec](#codec). 
- To use an S3 bucket, configure the `s3` source to use [Amazon S3 Select](https://docs.aws.amazon.com/AmazonS3/latest/userguide/selecting-content-from-objects.html) instead of Data Prepper to parse S3 objects.
```

`codec` | Yes | Codec  | The [codec](#codec) to apply. 
`sqs` | Yes | SQS | The SQS configuration. See [sqs](#sqs) for more information.
`aws` | Yes | AWS | The AWS configuration. See [aws](#aws) for more information.
`on_error` | No | String | Determines how to handle errors in Amazon SQS. Can be either `retain_messages` or `delete_messages`. `retain_messages` leaves the message in the Amazon SQS queue and tries to send the message again. This is recommended for dead-letter queues. `delete_messages` deletes failed messages. Default is `retain_messages`.
`buffer_timeout` | No | Duration  | The amount of time allowed for writing events to the Data Prepper buffer before timeout occurs. Any events that the Amazon S3 source cannot write to the buffer during the set amount of time are discarded. Default is `10s`.
`records_to_accumulate` | No | Integer  | The number of messages that accumulate before being written to the buffer. Default is `100`.
`metadata_root_key` | No | String | The base key for adding S3 metadata to each event. The metadata includes the key and bucket for each S3 object. Default is `s3/`.

> `default_bucket_owner` | No | String | The AWS account Id for the owner of S3 buckets. See [Cross-account S3 access](#s3_bucket_ownership) for more details.

Reviewer: ```suggestion
`default_bucket_owner` | No | String | The AWS account ID for the owner of S3 buckets. For more information, see [Cross-account S3 access](#s3_bucket_ownership)
```

> `bucket_owners` | No | Map  | A map of bucket names to owning account Ids. See [Cross-account S3 access](#s3_bucket_ownership) for more details.

Reviewer: ```suggestion
`bucket_owners` | No | Map  | A map of bucket names that includes the IDs of the accounts that own that buckets. For more information, see [Cross-account S3 access](#s3_bucket_ownership) for more details.
```

`notification_source` | No | String | Determines how notifications are received by SQS. Must be `s3` or `eventbridge`. `s3` represents notifications that are directly sent from Amazon S3 to Amazon SQS or fanout notifications from Amazon S3 to Amazon Simple Notification Service (Amazon SNS) to Amazon SQS. `eventbridge` represents notifications from [Amazon EventBridge](https://aws.amazon.com/eventbridge/) and [Amazon Security Lake](https://aws.amazon.com/security-lake/). Default is `s3`. 
`compression` | No | String  | The compression algorithm to apply: `none`, `gzip`, `snappy`, or `automatic`. Default is `none`.
`codec` | Yes | Codec  | The [codec](#codec) to apply. 
`sqs` | Yes | SQS | The SQS configuration. See [sqs](#sqs) for more information.
`aws` | Yes | AWS | The AWS configuration. See [aws](#aws) for more information.
`on_error` | No | String | Determines how to handle errors in Amazon SQS. Can be either `retain_messages` or `delete_messages`. `retain_messages` leaves the message in the Amazon SQS queue and tries to send the message again. This is recommended for dead-letter queues. `delete_messages` deletes failed messages. Default is `retain_messages`.

> `buffer_timeout` | No | Duration  | The amount of time allowed for writing events to the Data Prepper buffer before timeout occurs. Any events that the Amazon S3 source cannot write to the buffer during the amount of time set are discarded. Default is `10s`.

Reviewer 2: ```suggestion
`buffer_timeout` | No | Duration  | The amount of time allowed for writing events to the Data Prepper buffer before timeout occurs. Any events that the Amazon S3 source cannot write to the buffer during the specified amount of time are discarded. Default is `10s`.
```

`records_to_accumulate` | No | Integer  | The number of messages that accumulate before being written to the buffer. Default is `100`.
`metadata_root_key` | No | String | The base key for adding S3 metadata to each event. The metadata includes the key and bucket for each S3 object. Default is `s3/`.

> `default_bucket_owner` | No | String | The AWS account ID for the owner of S3 buckets. For more information, see [Cross-account S3 access](#s3_bucket_ownership)

Reviewer 2: ```suggestion
`default_bucket_owner` | No | String | The AWS account ID for the owner of an S3 bucket. For more information, see [Cross-account S3 access](#s3_bucket_ownership).
```

> `bucket_owners` | No | Map  | A map of bucket names that includes the IDs of the accounts that own that buckets. For more information, see [Cross-account S3 access](#s3_bucket_ownership) for more details.

Reviewer 2: ```suggestion
`bucket_owners` | No | Map  | A map of bucket names that includes the IDs of the accounts that own the buckets. For more information, see [Cross-account S3 access](#s3_bucket_ownership).
```

`visibility_timeout` | No | Duration | The visibility timeout to apply to messages read from the Amazon SQS queue. This should be set to the amount of time that Data Prepper may take to read all the S3 objects in a batch. Default is `30s`.
`wait_time` | No | Duration | The amount of time to wait for long polling on the Amazon SQS API. Default is `20s`.
`poll_delay` | No | Duration | A delay placed between the reading and processing of a batch of Amazon SQS messages and making a subsequent request. Default is `0s`.

> `visibility_duplication_protection` | No | Boolean | If set to `true`, Data Prepper attempts to avoid duplicate processing by extending the visibility timeout of SQS messages. Until the data reaches the sink, Data Prepper will regularly call `ChangeMessageVisibility` to avoid reading the S3 object again. To use this feature, you need to grant permissions to `sqs:ChangeMessageVisibility` on the IAM role. Default is `false`.

Reviewer 2: ```suggestion
`visibility_duplication_protection` | No | Boolean | If set to `true`, Data Prepper attempts to avoid duplicate processing by extending the visibility timeout of SQS messages. Until the data reaches the sink, Data Prepper will regularly call `ChangeMessageVisibility` to avoid rereading of the S3 object. To use this feature, you need to grant permissions to `sqs:ChangeMessageVisibility` on the IAM role. Default is `false`.
```

`region` | No | String | The AWS Region to use for credentials. Defaults to [standard SDK behavior to determine the Region](https://docs.aws.amazon.com/sdk-for-java/latest/developer-guide/region-selection.html).
`sts_role_arn` | No | String | The AWS Security Token Service (AWS STS) role to assume for requests to Amazon SQS and Amazon S3. Defaults to `null`, which will use the [standard SDK behavior for credentials](https://docs.aws.amazon.com/sdk-for-java/latest/developer-guide/credentials.html).
`aws_sts_header_overrides` | No | Map | A map of header overrides that the IAM role assumes for the sink plugin.

> `sts_external_id` | No | String | An STS external Id to use when Data Prepper assumes the role. See the documentation for ExternalId in the [STS AssumeRole](https://docs.aws.amazon.com/STS/latest/APIReference/API_AssumeRole.html) reference for more details.

Reviewer: ```suggestion
`sts_external_id` | No | String | An STS external ID to use when Data Prepper assumes the STS role. For more information, see the documentation for `ExternalID` in the [STS AssumeRole](https://docs.aws.amazon.com/STS/latest/APIReference/API_AssumeRole.html) reference.
```

`region` | No | String | The AWS Region to use for credentials. Defaults to [standard SDK behavior to determine the Region](https://docs.aws.amazon.com/sdk-for-java/latest/developer-guide/region-selection.html).
`sts_role_arn` | No | String | The AWS Security Token Service (AWS STS) role to assume for requests to Amazon SQS and Amazon S3. Defaults to `null`, which will use the [standard SDK behavior for credentials](https://docs.aws.amazon.com/sdk-for-java/latest/developer-guide/credentials.html).
`aws_sts_header_overrides` | No | Map | A map of header overrides that the IAM role assumes for the sink plugin.

> `sts_external_id` | No | String | An STS external ID to use when Data Prepper assumes the STS role. For more information, see the documentation for `ExternalID` in the [STS AssumeRole](https://docs.aws.amazon.com/STS/latest/APIReference/API_AssumeRole.html) reference.

Reviewer 2: ```suggestion
`sts_external_id` | No | String | An STS external ID used when Data Prepper assumes the STS role. For more information, see the `ExternalID` documentation in the [STS AssumeRole](https://docs.aws.amazon.com/STS/latest/APIReference/API_AssumeRole.html) API reference.
```

`end_time` | No | String  | The time after which no objects will be scanned after the given `end_time`. This should follow [ISO LocalDateTime](https://docs.oracle.com/javase/8/docs/api/java/time/format/DateTimeFormatter.html#ISO_LOCAL_DATE_TIME) format, for example, `023-01-23T10:00:00`. If `start_time` is configured along with `end_time`, all objects after `start_time` and before `end_time` will be processed. `end_time` and `range` cannot be used together. 
`range` | No  | String | The time range from which objects are scanned from all buckets. Supports ISO_8601 notation strings, such as `PT20.345S` or `PT15M`, and notation strings for seconds (`60s`) and milliseconds (`1600ms`). `start_time` and `end_time` cannot be used with `range`.  Range `P12H` scans all the objects modified in the last 12 hours from the time pipeline started. 
`buckets` | Yes  | List | A list of [scan buckets](#scan_bucket) to scan.
`scheduling` | No  | List | The configuration for scheduling periodic scans on all buckets. `start_time`, `end_time` and `range` can not be used if scheduling is configured. 

> ### scan bucket<a name="scan_bucket"></a>

Reviewer: We don't need the `a name` here:

```suggestion
### scan bucket
```

`range` | No  | String | The time range from which objects are scanned from all buckets. Supports ISO_8601 notation strings, such as `PT20.345S` or `PT15M`, and notation strings for seconds (`60s`) and milliseconds (`1600ms`). `start_time` and `end_time` cannot be used with `range`.  Range `P12H` scans all the objects modified in the last 12 hours from the time pipeline started. 
`buckets` | Yes  | List | A list of [scan buckets](#scan-bucket) to scan.
`scheduling` | No  | List | The configuration for scheduling periodic scans on all buckets. `start_time`, `end_time` and `range` can not be used if scheduling is configured. 

<!-- vale off -->
### scan bucket
<!-- vale on -->

Option | Required | Type | Description
:--- | :--- |:-----| :---
`bucket` | Yes | Map  | Provides options for each bucket.

> You can configure the following options inside the `bucket` setting.

Reviewer 2: "in" instead of "inside"?

> You can configure the following options inside the `bucket` setting.

Reviewer: ```suggestion
You can configure the following options in the `bucket` setting map.
```

`end_time` | No | String  | The time after which no objects will be scanned after the given `end_time`. This should follow [ISO LocalDateTime](https://docs.oracle.com/javase/8/docs/api/java/time/format/DateTimeFormatter.html#ISO_LOCAL_DATE_TIME) format, for example, `023-01-23T10:00:00`. If `start_time` is configured along with `end_time`, all objects after `start_time` and before `end_time` will be processed. `end_time` and `range` cannot be used together. 
`range` | No  | String | The time range from which objects are scanned from all buckets. Supports ISO_8601 notation strings, such as `PT20.345S` or `PT15M`, and notation strings for seconds (`60s`) and milliseconds (`1600ms`). `start_time` and `end_time` cannot be used with `range`.  Range `P12H` scans all the objects modified in the last 12 hours from the time pipeline started. 

> `buckets` | Yes  | List | A list of [scan buckets](#scan_bucket) to scan.

Reviewer: ```suggestion
`buckets` | Yes  | List | A list of [scan buckets](#scan-bucket) to scan.
```

* `s3ObjectsNotFound`: The number of S3 objects that the `s3` source failed to read due to an S3 "Not Found" error. These are also counted toward `s3ObjectsFailed`.
* `s3ObjectsAccessDenied`: The number of S3 objects that the `s3` source failed to read due to an "Access Denied" or "Forbidden" error. These are also counted toward `s3ObjectsFailed`. 
* `s3ObjectsSucceeded`: The number of S3 objects that the `s3` source successfully read.

> * `s3ObjectNoRecordsFound`: The number of S3 objects that resulted in 0 records added to the buffer by the `s3` source.

Reviewer 2: ```suggestion
* `s3ObjectNoRecordsFound`: The number of S3 objects that resulted in 0 records being added to the buffer by the `s3` source.
```

> * `s3ObjectNoRecordsFound` -- The number of S3 objects that resulted in 0 records added to the buffer by the `s3` source.

Reviewer: ```suggestion
* `s3ObjectNoRecordsFound`: The number of S3 objects that resulted in 0 records added to the buffer by the `s3` source.
```

> * `s3ObjectsDeleted` -- The number of S3 objects deleted by the `s3` source.

Reviewer: ```suggestion
* `s3ObjectsDeleted`: The number of S3 objects deleted by the `s3` source.
```

> * `s3ObjectsDeleteFailed` -- The number of S3 objects that the `s3` source failed to delete.

Reviewer: ```suggestion
* `s3ObjectsDeleteFailed`: The number of S3 objects that the `s3` source failed to delete.
```

> * `s3ObjectsEmpty` -- The number of S3 objects that are empty by having a size of `0`. These will be skipped.

Reviewer: ```suggestion
* `s3ObjectsEmpty`: The number of S3 objects that are considered empty since they have a size of `0`. These objects will be skipped by the `s3` source.
```

> * `s3ObjectsEmpty`: The number of S3 objects that are considered empty since they have a size of `0`. These objects will be skipped by the `s3` source.

Reviewer 2: ```suggestion
* `s3ObjectsEmpty`: The number of S3 objects that are considered empty because they have a size of `0`. These objects will be skipped by the `s3` source.
```

* `sqsMessagesReceived`: The number of Amazon SQS messages received from the queue by the `s3` source.
* `sqsMessagesDeleted`: The number of Amazon SQS messages deleted from the queue by the `s3` source.
* `sqsMessagesFailed`: The number of Amazon SQS messages that the `s3` source failed to parse.
* `sqsMessagesDeleteFailed` -- The number of SQS messages that the `s3` source failed to delete from the SQS queue.

> * `sqsVisibilityTimeoutChangedCount` -- The number of times that the `s3` source has changed the visibility timeout for an SQS message. This includes multiple visibility timeout changes on the same message.

Reviewer: ```suggestion
* `sqsVisibilityTimeoutChangedCount`: The number of times that the `s3` source changed the visibility timeout for an SQS message. This includes multiple visibility timeout changes on the same message.
```

> * `sqsVisibilityTimeoutChangeFailedCount` -- The number of times that the `s3` source failed to change the visibility timeout for an SQS message. This includes multiple visibility timeout change failures on the same message.

Reviewer: ```suggestion
* `sqsVisibilityTimeoutChangeFailedCount`: The number of times that the `s3` source failed to change the visibility timeout for an SQS message. This includes multiple visibility timeout change failures on the same message.
```

> * `acknowledgementSetCallbackCounter` -- The number of times that the `s3` source receives an acknowledgment from Data Prepper acknowledgments.

Reviewer: ```suggestion
* `acknowledgementSetCallbackCounter`: The number of times that the `s3` source receives an acknowledgment from Data Prepper.
```

Reviewer: Let's make sure that this change retains the technical meaning.

> * `acknowledgementSetCallbackCounter`: The number of times that the `s3` source receives an acknowledgment from Data Prepper.

Reviewer 2: ```suggestion
* `acknowledgementSetCallbackCounter`: The number of times that the `s3` source received an acknowledgment from Data Prepper.
```
